### PR TITLE
Feature/441 move project memberships to separate table

### LINKED ISSUE #441: Move project memberships to a separate table
- [x]  This table (project_memberships) will store the level of access to each project for the users
- project_id
- user_id
- level (owner, editor, reader)

- [x] Remove the "user_id" field from the projects table. Create "owner" records for existing projects before that.

- [x] Change the authorization checks in the controllers to use the new table

### PR DESCRIPTION
**Pull request by lmatayoshi**
Fixes #441

### CODE REVIEW

File: web/helpers/user_helper.ex [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
 
   def current_user(conn) do
     conn.assigns.current_user
   end
 
   def authorize(project, conn) do
-    if project.user_id != current_user(conn).id do
+    user_id = current_user(conn).id
+    memberships = project |> assoc(:project_memberships) |> where([m], m.user_id == ^user_id) |> Repo.all
+    if !Enum.any?(memberships, fn m -> m.level == "owner" or m.level == "editor" end) do
```
Comment (mverzilli) [inline web/helpers/user_helper.ex]: Why not include the level as a clause to the `where` in the query?

File: web/views/project_view.ex [hunk 2]
```diff
@@ -19,7 +19,7 @@ defmodule Ask.ProjectView do
 
   defp render_one(project) do
     %{id: project.id,
-      user_id: project.user_id,
+      # user_id: project.user_id,
```
Comment (mverzilli) [inline web/views/project_view.ex]: Maybe remove this line if it is commented?

File: priv/repo/migrations/20161220152839_remove_user_id_from_projects.exs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+defmodule Ask.Repo.Migrations.RemoveUserIdFromProjects do
+  use Ecto.Migration
+
+  def change do
+    Ask.Repo.transaction fn ->
+      Ask.Repo.query("ALTER TABLE projects DROP FOREIGN KEY projects_user_id_fkey, DROP COLUMN user_id")
```
Comment (waj) [inline priv/repo/migrations/20161220152839_remove_user_id_from_projects.exs]: Why not use something like:
```elixir
alter table(:projects) do
  remove :user_id
end
```

File: priv/repo/migrations/20161220144658_insert_project_memberships.exs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+defmodule Ask.Repo.Migrations.InsertProjectMemberships do
+  use Ecto.Migration
+  alias Ask.Repo
+  alias Ask.Project
+  alias Ask.ProjectMembership
+
+  def change do
+    Ask.Repo.transaction fn ->
+      Project |> Repo.all |> Enum.each(fn p ->
+        changeset = ProjectMembership.changeset(%ProjectMembership{}, %{project_id: p.id, user_id: p.user_id, level: "owner"})
```
Comment (waj) [inline priv/repo/migrations/20161220144658_insert_project_memberships.exs]: Using models for the migrations is not recommended

File: test/controllers/questionnaire_controller_test.exs [hunk 5]
```diff
@@ -117,15 +118,15 @@ defmodule Ask.QuestionnaireControllerTest do
 
     test "updates project updated_at when questionnaire is created", %{conn: conn, user: user}  do
       datetime = Ecto.DateTime.cast!("2000-01-01 00:00:00")
-      project = insert(:project, user: user, updated_at: datetime)
+      project = create_project_for_user(user)
```
Comment (asterite) [inline test/controllers/questionnaire_controller_test.exs]: Was this `updated_at: datetime` used in the test? It's strange that this doesn't make travis fail... maybe the test wasn't correct after all.

File: web/models/user.ex [hunk 6]
```diff
@@ -5,9 +5,11 @@ defmodule Ask.User do
     field :email, :string
     field :encrypted_password, :string
 
-    has_many :projects, Ask.Project
+    # has_many :projects, Ask.Project
```
Comment (asterite) [inline web/models/user.ex]: I would remove this comment

File: test/controllers/questionnaire_controller_test.exs (hunk 5 above)
Comment (lmatayoshi) [inline test/controllers/questionnaire_controller_test.exs]: Spec was correct, I've already reverted changes

**Pull request merged**